[로또 미션] 김명지 미션 제출합니다.

Starlight258 opened the pull request:
안녕하세요 진영님 !
로또 3단계 미션을 완료했습니다 😁

## 궁금한 점
- 댓글에 남겼습니다!

✅ 구현된 코드에 대한 설명은 `README.md`에서 보실 수 있습니다 ! 

리뷰 부탁드립니다 !! 감사합니다~ 🙇‍♀️

Starlight258 committed:
refactor: 리팩토링

- 복잡한 연산을 메서드로 분리한다.
- 매직넘버는 상수로 분리한다.
- 한번만 사용되는 변수는 제거한다.

Starlight258 committed:
feat: 3단계 1차 구현

Starlight258 committed:
feat: enum 도입

- LottoMatch : 당첨 금액 및 당첨 결과가 저장된 key가 저장된 enum class
- LottoBoundary: 로또 경계값 enum class

Starlight258 committed:
fix: 2등과 3등이 함께 세어지는 에러 해결

- 로또 번호가 5개 일치하고 보너스 번호까지 일치할 경우 2등으로만 세어지도록 변경한다.

Starlight258 committed:
test: 변경된 코드에 따라 테스트 추가

Starlight258 committed:
rename: 상수 패키지로 분리

- 상수 클래스는 constant 패키지로 분리한다.

Starlight258 committed:
fix: 매칭 결과 key로 조회하도록 수정

- 매칭 결과는 cnt가 아닌 key로 조회해야한다. 2등과 3등은 매칭 결과는 같지만, 보너스볼 일치 여부가 다르므로 key가 다르다.

Starlight258 committed:
refactor: 리팩토링

- 복잡한 연산 메서드로 분리한다.
- 주생성자를 호출하여 부생성자를 둔다.

Starlight258 committed:
refactor: 메서드 간결하게 수정

Starlight258 committed:
style: formatting

Starlight258 committed:
docs: README 업데이트

Starlight258 commented on src/main/java/org/duckstudy/model/lotto/Lottos.java:
```diff
@@ -26,9 +25,14 @@ public static Lottos generateLottosByPrice(Price price) {
-        return new LottoResult(lottos.stream()
-                .collect(groupingBy(winningLotto::countMatchingNumber, summingInt(e -> 1))));
+    public LottoResult calculateWinningResult(Lotto winningLotto, LottoNumber bonusNumber) {
+        LottoResult lottoResult = new LottoResult(Map.of());
+        for (Lotto lotto : lottos) {
+            lottoResult = lottoResult.updateResult(
+                    lotto.countMatchingNumber(winningLotto),
+                    lotto.matchBonusNumber(bonusNumber));
+        }
+        return lottoResult;
```
기존 로또 로직에 보너스 번호가 추가되면서 로직이 살짝 복잡해졌습니다.
stream으로 구현 코드를 수정해봤는데, 굉장히 복잡해지더라구요..!
전에 이야기하셨던것처럼, stream은 행위를 잘 드러나게 작성하는게 중요한데 어느정도 복잡도선에서 stream을 사용하면 좋을까요?

jinyoungchoi95 replied:
어떻게 복잡해지는지 잘 모르겠어요. 말씀해주신대로 stream에서 행위가 잘 드러나지 않으면 stream을 활용할 이유가 줄어드는 건 맞아요.

복잡하지 않도록 로직을 구성하는 방법은 없을까요?

Starlight258 replied:
그렇군요! 결과 업데이트하는 부분을 메서드로 분리해서 행위가 잘 드러나게 수정했습니다!
```java
   public LottoResult calculateTotalLottoResult(Lotto winningLotto, LottoNumber bonusNumber) {
        return lottos.stream()
                .map(lotto -> calculateLottoResult(lotto, winningLotto, bonusNumber, new LottoResult(Map.of())))
                .reduce(new LottoResult(Map.of()), LottoResult::merge);
    }

    private LottoResult calculateLottoResult(Lotto lotto, Lotto winningLotto, LottoNumber bonusNumber, LottoResult result
    ) {
        int matchingCount = lotto.countMatchingNumber(winningLotto);
        boolean matchBonus = lotto.containsNumber(bonusNumber);
        return result.updateResult(matchingCount, matchBonus);
    }
```

jinyoungchoi95 commented on src/main/java/org/duckstudy/view/InputView.java:
```diff
@@ -27,12 +27,26 @@ public int[] inputWinningLotto() {
         }
     }
+
+    private int[] inputLottoNumber() throws IOException {
```
> 배열 대신 컬렉션을 사용한다.

Starlight258 replied:
요구사항에 맞게 수정하겠습니다! 감사합니다!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/constant/LottoMatch.java:
```diff
@@ -0,0 +1,32 @@
+
+public enum LottoMatch {
+
+    MIN_MATCH(3, 5000, 3),
+    MATCH_4(4, 50000, 4),
+    MATCH_5(5, 1500000, 5),
+    MATCH_5_WITH_BONUS(5, 30000000, -5),
+    MAX_MATCH(6, 2000000000, 6);
```
Match라는 표현보다는 1등부터 4등, 미당첨을 표현하는건 어떨까요? match는 로또라는 도메인을 담기보다는 그냥 일치 갯수를 나타내는걸로보이네요.

Starlight258 replied:
로또 도메인이므로 등수로 표현하는 것이 더 명확하게 의미가 드러나네요! 
상수명은 도메인에서 가지는 의미가 담기도록 작성하겠습니다! 감사합니다~

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/constant/LottoBoundary.java:
```diff
@@ -0,0 +1,19 @@
+package org.duckstudy.model.lotto.constant;
+
+public enum LottoBoundary {
+
+    START_INCLUSIVE_NUMBER(1),
+    END_INCLUSIVE_NUMBER(45),
+    LOTTO_SIZE(6),
+    LOTTO_PRICE(1000);
```
상수 클래스처럼 사용하셨는데 각 객체가 본인의 의미있는 도메인정보를 들고 있는 것이 아닌 이런 클래스가 객체지향과 멀어지는 것 아닐까요?

형태는 어쩌다보니 각 상수와 value가 있어 이런 형태가 되었지만 서로 연관없는 enum값들을 함께 사용하는 것이 enum의 목적이 맞을까요?

Starlight258 replied:
맞습니다, enum은 연관된 상수들의 집합입니다 ! 코드 작성시에 상수끼리의 관계는 크게 고려하지 못했네요 !
enum이 서로 관련된 상수를 그룹화하고, 사용되는 맥락을 내포할 수 있다는 사실을 몰랐네요!

`START_INCLUSIVE_NUMBER,` `END_INCLUSIVE_NUMBER`는 로또 번호 범위를 나타내는 enum 클래스로 만들고, `LOTTO_SIZE`와  `LOTTO_PRICE`는 각각 Lotto와 Price 클래스에서만 사용하므로 클래스 내 상수로 정의하겠습니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/LottoNumber.java:
```diff
@@ -1,17 +1,18 @@
 
-    private static final int START_INCLUSIVE_NUMBER = 1;
-    private static final int END_INCLUSIVE_NUMBER = 45;
     private static final LottoNumber[] cache;
```
> 배열 대신 컬렉션을 사용한다.

Starlight258 replied:
수정하겠습니다~ 감사합니다!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/Price.java:
```diff
@@ -16,25 +14,21 @@ public Price(int price) {
     private void validatePrice(int price) {
-        if (price < INCLUSIVE_MIN_PRICE) {
-            throw new IllegalArgumentException(String.format("가격은 %d원 이상이어야 합니다.", INCLUSIVE_MIN_PRICE));
+        if (price < 0) {
```
다시 매직넘버로 변경된 이유가 있을까요?

Starlight258 replied:
0이 코드에서 자주 나타나다보니 상수로 모든 0을 정의하는 것이 좋을지 생각이 들었습니다.
그래서 다시 0의 값을 가진 상수들을 0으로 바꾸었습니다.!
그런데 지금 생각해보니 해당 메서드의 0은 가격의 초기값을 나타내니 상수로 정의해두는 것이 좋겠네요!

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/LottoResult.java:
```diff
@@ -21,16 +19,32 @@ public LottoResult(Map<Integer, Integer> result) {
+        return result.getOrDefault(count, 0);
+    }
+
+    public LottoResult updateResult(int matchingCount, boolean matchBonusNumber) {
```
update를 통해 `Map<Integer, Integer> result`의 값을 가변하게 사용하는 것이 아닌 생성 시점에서 불변하게 필요한 값들을 넣을 수 있는 설계를 한번 고민해보시면 좋을 것 같아요.

Starlight258 replied:
merge 메서드를 만들어서 새로 생성된 LottoResult를 기존의 LottoResult와 병합하게 작성해서 불변성을 유지하도록 수정했습니다!

jinyoungchoi95 reviewed (changes requested):
안녕하세요 명지님 😃

몇가지 코멘트남겨두었으니 확인부탁드려요.

감사합니다🙇‍♂️

Starlight258 committed:
refactor: 메서드를 분리하여 메서드 행위 잘 드러나도록 수정

- 복잡한 연산을 메서드로 분리하여 stream 사용시 행위가 잘 드러나도록 수정한다.

Starlight258 committed:
refactor: 배열 대신 컬렉션을 사용하도록 수정한다.

Starlight258 committed:
style: code formatting

Starlight258 committed:
refactor: enum 클래스 목적에 맞게 리팩토링

- LottoNumberRange: 로또 번호 범위 상수 클래스
- LottoRank: 로또 도메인이 드러나도록 등수로 상수명 변경
- LOTTO_SIZE와 LOTTO_PRICE는 클래스 안에서만 사용하므로 각 클래스 내의 상수로 정의

Starlight258 committed:
refactor: 특별한 값을 가지는 0은 상수로 정의

- 가격의 초기값을 나타내는 0은 특별한 의미를 가지므로 상수로 정의한다.

Starlight258 committed:
feat: LottoResult에 merge 메서드 추가하여 불변성 유지

- LottoResult끼리 merge하는 메서드를 생성하여 LottoResult 업데이트시 result를 직접 수정하지 않고 새로운 LottoResult를 생성하여 불변성을 유지한다.

Starlight258 committed:
feat: 로또 번호 일치 개수와 보너스 번호 일치에 따른 당첨 등수 찾는 기능 구현

Starlight258 committed:
test: LottoTest 테스트 추가 및 리팩토링

Starlight258 committed:
refactor: 로또 수익률 계산 메서드 Price로 이동

- 로또 수익률 계산은 Price의 책임에 가까우므로 LottoResult에서 Price로 이동

Starlight258 committed:
test: 테스트 추가

Starlight258 committed:
refactor: 리팩토링

- OutputView의 내부 구현에 의존하지 않고 객체에게 메세지를 보내도록 수정
- 메서드명 변경

Starlight258 committed:
refactor: controller에서 Lotto 로직 분리

- 당첨 번호 중복 로직은 Lotto에 포함되도록 수정
- 메서드명 변경

Starlight258 committed:
refactor: 검증 메서드 분리

Starlight258 committed:
test: 테스트 코드 변경

- 변경된 코드에 따른 테스트 추가
- 메서드명 변경

Starlight258 commented:
@jinyoungchoi95
리뷰해주신 사항들 덕분에 많이 배웠습니다 🙇‍♀️
감사합니다 ! 😁

Starlight258 committed:
refactor: 10줄 넘지 않도록 메서드 수정

Starlight258 committed:
refactor: 리팩토링

- 매직넘버를 상수로 변경한다.
- 구매금액이 0원일때 다시 입력받도록 수정한다.

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/Price.java:
```diff
@@ -15,7 +18,7 @@ public Price(int price) {
     }
 
-    public static Price zero() {
+    public static Price initialize() {
```
`initialize`는 의미상 전달되는 것은 "초기화"에 가까워요. 초기값이라기보다는 무언가 변경된 것을 초기화한다는 것에 더 가깝기 때문에 다른 네이밍이 어떨까요? (기존의 zero나 차라리 initialPrice가 좀더 명확하지 않을까싶네요)

Starlight258 replied:
그렇네요! 값을 초기 상태로 되돌리는 것이 아닌 초기값으로 객체를 생성하는 것이니 다른 메서드명이 좋겠네요! 
수정하도록 하겠습니다~

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/constant/LottoNumberRange.java:
```diff
@@ -0,0 +1,17 @@
+
+public enum LottoNumberRange {
+
+    START_INCLUSIVE_NUMBER(1),
+    END_INCLUSIVE_NUMBER(45);
```
이런 값들도 사실 `LottoNumber`가 가지고 있어야 응집도가 더 올라갈 것 같아요.

상수를 public으로 열어두어야하는 것이 고민이라면 설계상에서 private으로 막아놓고 설계를 하는 방법을 고려해볼 수 있고, 혹은 상수를 public하게 여는 방법밖에 없으면 차라리 해당 클래스가 상수를 들고 있는게 낫다고 생각해요.

Starlight258 replied:
그렇군요..! 
굳이 `enum` 클래스를 생성하는 것보다, 상수가 `LottoNumber`와 밀접한 관련이 있다면 도메인에서 상수를 가지고 있는게 더 낫겠네요 ! 감사합니다 :)

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/constant/WinningRank.java:
```diff
@@ -0,0 +1,45 @@
+public enum WinningRank {
+
+    NONE(0, 0, 0),
+    FIFTH(3, 5000, 3),
```
```suggestion
    FIFTH(3, 5_000, 3),
```
3자리가 넘어갈때는 `_`로 가독성을 올릴 수 있어요.

Starlight258 replied:
오 그렇군요! 😀 반영하겠습니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/constant/WinningRank.java:
```diff
@@ -0,0 +1,45 @@
+        if (matchCount == SECOND.getMatchCount() && matchBonus) {
+            return SECOND;
+        }
+        return Stream.of(values())
+                .filter(winningLank -> winningLank.getMatchCount() == matchCount)
+                .findFirst()
+                .orElse(NONE);
```
filter에 있는 행위를 추상화해보면 "matchCount가 해당 WinnintRank.matchCount"와 일치하느냐에요. 그렇다면 NONE은 모든 값에 대해서 불일치하다기보다는 0, 1, 2 각각이 모두 filter에 들어갈 수 있는 행위가 되겠죠.

그래서 orElse로 1, 2에 대한 비즈니스를 처리하기보다는 `WinningRank`가 이를 처리할 수 있는 방법으로 처리하는게 도메인 전달에 좀 더 명확하지 않을까 싶어요. list를 활용한다거나, filter자체를 생각해본다면 functional interface를 활용해볼 수 있겠네요.

Starlight258 replied:
그렇네요! 매칭되지 않은 경우 NONE을 반환하는데, enum의 values에서 이미 NONE 여부를 확인하므로 예외 상황을 NONE으로 처리하는 것은 적절하지 않네요! 
WinningRank가 매칭 여부를 처리하고, 예외 상황의 경우 에러를 발생시키도록 수정하겠습니다.
함수형 인터페이스도 사용하겠습니다 !

```java

    NONE(0, 0, 0, (matchCount, matchBonus) -> matchCount < 3),
    FIFTH(3, 5_000, 3, (matchCount, matchBonus) -> matchCount == 3),
    ...

    private final BiPredicate<Integer, Boolean> isMatchPredicate;

    public static WinningRank findByMatchCountAndBonus(int matchCount, boolean matchBonus) {
        return Stream.of(values())
                .filter(winningLank -> winningLank.isMatch(matchCount, matchBonus))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("적절한 당첨 등수를 찾을 수 없습니다."));
    }

    public boolean isMatch(int matchCount, boolean matchBonus) {
        return isMatchPredicate.test(matchCount, matchBonus);
    }
```

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/constant/WinningRank.java:
```diff
@@ -0,0 +1,45 @@
+        this.key = key;
+    }
+
+    public static WinningRank findByMatchCountAndBonus(int matchCount, boolean matchBonus) {
```
이런 메소드는 테스트가 필요한 로직들이 많이 있어보여요.

Starlight258 replied:
enum도 테스트 범위에 포함되는군요! 테스트 추가하도록 하겠습니다 !

jinyoungchoi95 commented on src/main/java/org/duckstudy/model/lotto/LottoResult.java:
```diff
@@ -1,36 +1,44 @@
+        int matchingCount = lotto.countMatchingNumber(winningLotto);
+        boolean matchBonus = lotto.containsNumber(bonusNumber);
+
+        int key = WinningRank.findByMatchCountAndBonus(matchingCount, matchBonus)
+                .getKey();
```
`findByMatchCountAndBonus`의 반환인 `WinningRank`가 바로 사용되는 곳이 없다면 반환자체를 key를 반환하도록 하는건 어떨까싶어요.

Starlight258 replied:
key로만 사용되니 반환값을 key로 두는 것이 적절하겠네요 감사합니다!

jinyoungchoi95 reviewed (approved):
안녕하세요 명지님 😃

피드백 잘 반영해주셨네요. 다음 미션 진행을 위해 머지하겠습니다.

감사합니다🙇‍♂️

Starlight258 commented:
@jinyoungchoi95
리뷰 감사합니다 ! 😁
덕분에 많이 배웠습니다 🙇‍♀️
